hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/DataNodeMetricsService.java

```diff
@@ -0,0 +1,342 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
+import org.apache.hadoop.hdds.scm.server.OzoneStorageContainerManager;
+import org.apache.hadoop.hdds.server.http.HttpConfig;
+import org.apache.hadoop.ozone.recon.MetricsServiceProviderFactory;
+import org.apache.hadoop.ozone.recon.api.types.DataNodeMetricsServiceResponse;
+import org.apache.hadoop.ozone.recon.api.types.DatanodePendingDeletionMetrics;
+import org.apache.hadoop.ozone.recon.scm.ReconNodeManager;
+import org.apache.hadoop.ozone.recon.tasks.DataNodeMetricsCollectionTask;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Service for collecting and managing DataNode pending deletion metrics.
+ * Collects metrics asynchronously from all datanodes and provides aggregated results.
+ */
+@Singleton
+public class DataNodeMetricsService {
+  
+  private static final Logger LOG = LoggerFactory.getLogger(DataNodeMetricsService.class);
+  private static final int MAX_POOL_SIZE = 500;
+  private static final int KEEP_ALIVE_TIME = 60;
+  private static final int QUEUE_CAPACITY = 500;
+  private static final int POLL_INTERVAL_MS = 200;
+  private static final int PER_NODE_TIMEOUT_MS = 120000;
+
+  private final ThreadPoolExecutor executorService;
+  private final ReconNodeManager reconNodeManager;
+  private final boolean httpsEnabled;
+  private final int minimumApiDelayMs;
+  private final MetricsServiceProviderFactory metricsServiceProviderFactory;
+  
+  private MetricCollectionStatus currentStatus = MetricCollectionStatus.NOT_STARTED;
+  private List<DatanodePendingDeletionMetrics> pendingDeletionList;
+  private Long totalPendingDeletion = 0L;
+  private int totalNodesQueried;
+  private int totalNodesFailed;
+  private long lastCollectionEndTime;
+
+  @Inject
+  public DataNodeMetricsService(
+      OzoneStorageContainerManager reconSCM,
+      OzoneConfiguration config,
+      MetricsServiceProviderFactory metricsServiceProviderFactory) {
+    
+    this.reconNodeManager = (ReconNodeManager) reconSCM.getScmNodeManager();
+    this.httpsEnabled = HttpConfig.getHttpPolicy(config).isHttpsEnabled();
+    this.minimumApiDelayMs = (int) config.getTimeDuration(
+        OZONE_RECON_DN_METRICS_COLLECTION_MINIMUM_API_DELAY,
+        OZONE_RECON_DN_METRICS_COLLECTION_MINIMUM_API_DELAY_DEFAULT,
+        TimeUnit.MILLISECONDS);
+    this.metricsServiceProviderFactory = metricsServiceProviderFactory;
+    this.lastCollectionEndTime = 0;
```

Comment:
lastCollectionEndTime initial value is better to be -minimumApiDelayMs, otherwise, the first request to pending deletion API within the first minimumApiDelayMs seconds will return accepted but task is not started.